### DISCUSSION TITLE
Difficult color replacement

### DISCUSSION THREAD

**Discussion by jaka1-newbie** [ImageMagick/ImageMagick#4840] [Help]
Hi, experts! 
With ColorFuzz we can make another color with near colors.

But can we do difficult changes.
From one group color to another?
![from](https://user-images.githubusercontent.com/67605008/153774938-efbeea40-97c6-4e84-b00d-56b0b72ec6c4.png)
![to](https://user-images.githubusercontent.com/67605008/153774941-375180cf-8d1e-4970-bb81-e805f9de2c80.png)

**Comment by fmw42** [comment]
Use -modulate to brightness, saturation and/or hue.  See https://imagemagick.org/script/command-line-options.php#modulate and https://legacy.imagemagick.org/Usage/color_mods/#color_mods

**Comment by snibgo** [comment]
Another approach: the horse has a nearly constant hue (in HCL colorspace). So you can use that as a mask to do whatever change you want.

**Reply by jaka1-newbie** [reply]
> Another approach: the horse has a nearly constant hue (in HCL colorspace). So you can use that as a mask to do whatever change you want.

Can you show example of this miracle?

**Comment by snibgo** [comment]
Windows BAT script:
```
magick ^
  colHorse.png ^
  -alpha off ^
  -colorspace HCL ^
  -separate ^
  -delete 2 ^
  ( -clone 0 ^
    -evaluate AddModulus 58.5%% ^
    -solarize 50%% ^
    -evaluate Multiply 2 ^
    +write h.png ^
  ) ^
  ( -clone 1 ^
    -level 5,15%% ^
    +write c.png ^
  ) ^
  -delete 0-1 ^
  -compose Multiply -composite ^
  msk.png

magick ^
  colHorse.png ^
  ( +clone -modulate 100,100,30 +write k.png ) ^
  msk.png ^
  -compose Over -composite ^
  comp.png
```

The first command creates a mask, msk.png. It makes an image from the Hue channel of HCL. The horse hue is around 91.5%. To avoid continuity problems (99.9% looks very close to 0.1%, but the arithmetic has a large difference), we add 58.5% so the horse is now around 50%, and we solarize and multiply to make the horse at 100%, with other hues less than that.

The first command then makes an image from Chroma of HCL, and levels it so the horse (with high Chroma) becomes white and the background and bridle become black.

h.png and c.png are so we can see the individual effects of these selections. These lines can be removed.

We multiply these together to make the mask: white where the horse is, and black elsewhere.

We can use the mask in many ways. I show a modulation of hue. Compositing that over the original changes the horse but not the background.

comp.png is:
![comp](https://user-images.githubusercontent.com/33812368/153838130-9d58fcf3-ab5b-4c4a-8a14-7120d001d2f9.png)
